src/content/plugins.md

```diff
@@ -0,0 +1,329 @@
+programming way of iterating an input and combining the output into an accumulator. In our case we are merely 
+concatenating the output together into one big string.
+
+```js
+    build() {
+        const walkOptions = {
+            includeBasePath: true,
+            directories: false,
+            globs: this.fileMatchers,
+        };
+        
+        const content = this.inputPaths
+            .reduce((output, inputPath) => output + this.joinSeparator +
+                walkSync(inputPath, walkOptions)
+                    .map(file => fs.readFileSync(file, { encoding: 'UTF-8' }))
+                    .join(this.joinSeparator),
+            '');
+    
+        fs.writeFileSync(`${this.outputPath}/${this.outputFile}`, content);
+    }
+```
+
+Next we iterate each file the `inputPath` and read its contents. We do this via the map function, which 
+transforms the array of files into an array of the contents of each file. We then join all the files together
+with the separator character, which is returned as the result to the reduce method above. 
+
+After the above is complete, we now have all of the file contents within the `content` variable and all that is 
+left is to write that to the outputFile. 
+
+That's it, plugin complete. In our case, the plugin is entirely synchronous and as such just returns at the end.
+If we needed the plugin to be asynchronous, we could alternatively return a promise and Broccoli would wait until
+the promise resolved before continuing the build.
+
+Lastly, we export a default function that will be used when this plugin is imported into a build pipeline
+```js
+export default function concatPlugin(...params) {
+    return new ConcatPlugin(...params);
+}
+```
+
+E.g.
+
+```js
+// Brocfile.js
+import concat from 'concat-plugin';
+
+export default () => concat(['dir1', 'dir2']);
+```
+
+As you can see, there isn't really any magic happening here, it's all standard Node code, just wrapped in a 
+`build()` method that is provided an array of `inputPaths` and an `outputPath` to write to.
+
+# Caching
+
+Adding caching to a plugin is relatively straight forward. Caching allows a plugin to skip its `build()` method
+if it doesn't need to do anything because its inputs have no changed. We can use a couple of node packages to
+achieve this. [fs-tree-diff](https://www.npmjs.com/package/fs-tree-diff) is used to calculate a diff between the
+2 states of a directory, and [walk-sync](https://www.npmjs.com/package/walk-sync) which synchronously walks all
+the files in a directory and produces a list, which is passed to `fs-tree-diff`.
+
+Let's take a look:
+
+```js
+import Plugin from 'broccoli-plugin';
+import FSTree from 'fs-tree-diff';
+import walkSync from 'walk-sync';
+
+export class Cacher extends Plugin
+{
+  constructor (inputNodes, options) {
+    super(inputNodes, {
+      ...options,
+      persistentOutput: true,
+    });
+
+    this._previous = [];
+  }
+
+  _hasChanged() {
```

Comment:
@stefanpenner I feel like we could make a utility class in a node package to do this. Initialize it with `this.inputPaths` and let it do the rest, maintaining state etc and exposing a `hasChanges()` method that returns the changes since last call or something.

Reply:
@oligriffiths that could be good but might be worth mentioning it in an aside (“or use the function from node-package which does much the same thing)?

We may lose some of the impact of realizing how powerful these tools are if we’re always having to pull in more code from another package vs just rolling our own ...

Reply:
I think we could supply this as an additional import from `broccoli-plugin`, but it is also worth explaining how simple it is to do it yourself.